---
title: "Prpf8"
output: 
  bookdown::word_document2:
    toc: true
    number_sections: false
    reference_docx: "output_template.docx"
---

```{r setup, echo=FALSE, message=FALSE, warning = FALSE, results = "hide"}
knitr::opts_chunk$set(echo = FALSE, fig.width = 8, fig.height = 5, cache = TRUE)
library(tidyverse)
library(bayesplot)
library(tidybayes)
library(ggdist)
library(brms)
library(patchwork)
options(mc.cores = 4, brms.backend = "cmdstanr")
theme_set(cowplot::theme_cowplot())
devtools::load_all()

options(width = 160)

document_output <- isTRUE(getOption('knitr.in.progress'))

if(document_output) {
  table_format <- function(x, caption = NULL, align = NULL) { 
    knitr::kable(x, caption = caption, align = align) 
  }
} else {
  table_format <- function(x, caption = NULL, align = NULL) { x }
}

```

```{r}
f <- here::here("private_data/qPCR/qPCR Prpf8 d17 4w final.xlsx")
litter_sheet <- "Litter "
litters <- read_all_litters(f, litter_sheet, c("A2:I10", "A13:I18", "A21:I28", "K2:S8","K11:S21","K24:S34"))
```

```{r}
a1 <- read_qpcr(f, "qPCR set a1", range = "A5:AR31")
a2 <- read_qpcr(f, "qPCR set a2", range = "A5:Z22")
a3 <- read_qpcr(f, "qPCR set a3", range = "A5:AU22")
a4 <- read_qpcr(f, "qPCR set a4", range = "A5:AC56")

qpcr_d17 <- rbind(a1,a2,a3,a4)

```

# Exploratory plots


```{r hist-housekeeping, fig.cap = "Histogram of Cq numbers for the three housekeeping genes used. Gapdh seems to be the most consistent. wt/d17 - wildtype/d17 homozygotes, het - heterozygotes"}
housekeeping <- c("Ubb", "Gapdh", "Actb") 

qpcr_d17 %>% filter(primer %in% housekeeping) %>%
  ggplot(aes(x = Cq)) + geom_histogram(binwidth = 0.2) + facet_grid(genotype ~ primer, scales = "free")
```


```{r box-housekeeping, fig.cap = "Cq values of the housekeeping genes stratified by genotype and sex. Each dot is a single technical replicate."}
qpcr_d17 %>% filter(primer %in% housekeeping) %>%
  ggplot(aes(y = Cq, x = genotype)) + geom_boxplot(outlier.shape = NA) + 
  geom_point(aes(color = sex, shape = sex), position = position_jitter(width = 0.25), alpha = 0.7) + 
  facet_wrap( ~ primer, scales = "free")

```


```{r parcoord-housekeeping, fig.cap = "Average Cq number across technical replicates for the housekeeping genes in each biological replicate. Each line connects measurements from the same biological replicate."}
qpcr_d17 %>% filter(primer %in% housekeeping) %>%
  group_by(animal_plate, genotype, primer_short, sex) %>%
  summarise(Cq = mean(Cq), .groups = "drop") %>%
  ggplot(aes(y = Cq, x = primer_short, color = sex, group = animal_plate)) + 
  geom_line(alpha = 0.7) + 
  facet_wrap( ~ genotype, scales = "free")

```


The set of plots below shows Cq numbers for all the measured primers stratified by sex and genotype. Each dot is a single technical replicate.

```{r, fig.width=10, fig.height=6}
to_compare <- setdiff(unique(qpcr_d17$primer), housekeeping)
step_size <- 6
for(step in 1:ceiling(length(to_compare) / step_size)) {
  primers_to_show <- to_compare[((step - 1) * step_size + 1):(step * step_size)]
  # print(qpcr_d17 %>% filter(primer %in% primers_to_show) %>%
  #   ggplot(aes(x = ct)) + geom_histogram(binwidth = 0.2) + facet_grid(genotype ~ primer, scales = "free_y")
  # )
  print(qpcr_d17 %>% filter(primer %in% primers_to_show) %>%
    ggplot(aes(y = Cq, x = genotype)) + geom_boxplot(outlier.shape = NA) + 
  geom_point(aes(color = sex, shape = sex), position = position_jitter(width = 0.25), alpha = 0.7) + 
  facet_wrap( ~ primer_short, scales = "free")
  )

}

```

# Fitting the model

```{r}
#qpcr_d17_model <- qpcr_d17 
 qpcr_d17_model <- qpcr_d17 %>% mutate(primer_short = factor(primer_short),
                                       sex = factor(sex))

priors = c(prior(normal(0,1), class = "sd"),
           prior(normal(0,2), class = "b"))

# fit1 <- brm(Cq ~ primer_short * mo(genotype_ord) + (1 || run) + (primer_short || sex + animal_no + litter + animal_plate + animal_replicate), data = qpcr_d17, prior = priors)
fit1 <- brm(
  bf(Cq ~ primer_short * genotype + sex*primer_short + (1 || run + animal_no + animal_no:primer_short + litter:primer_short),
     sigma ~ (1 || primer_short + run)
     ), 
  data = qpcr_d17_model, 
  prior = priors, refresh = 100,
  file = here::here("local_temp_data/fit1_all"), file_refit = "on_change")
fit1
```

```{r}
pred_fit1 <- posterior_predict(fit1, cores = 1)
```

Posterior predictive checks to asses model fit.

```{r}
ppc_stat_grouped(qpcr_d17_model$Cq, pred_fit1, group = qpcr_d17_model$primer_short, stat = "sd")
ppc_stat_grouped(qpcr_d17_model$Cq, pred_fit1, group = qpcr_d17_model$genotype_ord, stat = "sd")
ppc_stat_grouped(qpcr_d17_model$Cq, pred_fit1, group = paste0(qpcr_d17_model$primer_short, qpcr_d17_model$genotype_ord), stat = "sd")
ppc_stat_grouped(qpcr_d17_model$Cq, pred_fit1, group = paste0(qpcr_d17_model$primer_short, qpcr_d17_model$genotype_ord), stat = "mean")
ppc_stat_grouped(qpcr_d17_model$Cq, pred_fit1, group = qpcr_d17_model$sex, stat = "sd")
ppc_stat_grouped(qpcr_d17_model$Cq, pred_fit1, group = qpcr_d17_model$run, stat = "sd")
ppc_stat_grouped(qpcr_d17_model$Cq, pred_fit1, group = qpcr_d17_model$run, stat = "mean")

```

# Comparisons

The basic quantity of interest are the ratios of two products, we treat one product as numerator and one as denominator. When detecting circRNAs we take the numerator to be the primer that should be specific to the circRNA and denominator as the primer that should correspond to the linear form, i.e. the larger ratio for circRNA, the higher proportion of the products are circRNA. For detecting alternative exons, we take the primer including the alternative exons as the numerator, i.e. the the larger ratio for alternative exon, the more products contain the alternative exon. 

Those are the ratios products we use:

```{r}
comparisons <- readxl::read_excel(here::here("private_data/qPCR/qPCR selected amplicons.xlsx"), sheet = "qPCR amplicons to be compared", range = "T2:W40") %>% filter(!is.na(label)) %>%
  mutate(numerator = primer_to_short(numerator)) %>%
  pivot_longer(cols = starts_with("denominator"), names_to = "variant", names_prefix = "denominator_",
               values_to = "denominator") %>%
  mutate(denominator = primer_to_short(denominator)) %>%
  filter(!is.na(denominator)) %>%
  group_by(label) %>%
  mutate(label_variant = if_else(rep(n() > 1, n()), paste0(label, " - ", variant), label)) %>%
  ungroup()

comparisons <- comparisons %>% 
  mutate(denominator = factor(denominator, levels = levels(qpcr_d17_model$primer_short)),
         numerator = factor(numerator, levels = levels(qpcr_d17_model$primer_short)))

if(any(is.na(comparisons$numerator)) || any(is.na(comparisons$denominator))) {
  stop("Bad factor levels")
}

comparisons %>% select(label, variant, numerator, denominator) %>% table_format()

```


```{r}
eff_variants <- data.frame(eff_numerator = c(1.5, 1.8, 2), eff_denominator = c(2, 1.8, 1.5), eff_label = factor(1:3, labels = c("Favor denom", "Neutral", "Favor num")))

comparisons_eff <- comparisons %>%
  crossing(eff_variants) %>%
  mutate(comparison_id = 1:n())

# pred_data_numerator_wt <- qpcr_d17_model %>% filter(primer_short %in% comparisons$numerator, genotype_ord == "wt",
#                                               sex == "F") %>% 
#   group_by(primer_short) %>% top_n(1)  
# 

pred_data_numerator_wt <- data.frame(primer_short = comparisons_eff$numerator,
                                     genotype = "wt",
                                     sex = "F"
                                     )
pred_data_numerator_d17 <- data.frame(primer_short = comparisons_eff$numerator,
                                     genotype = "d17",
                                     sex = "F"
                                     )
pred_data_denominator_wt <- data.frame(primer_short = comparisons_eff$denominator,
                                     genotype = "wt",
                                     sex = "F"
                                     )
pred_data_denominator_d17 <- data.frame(primer_short = comparisons_eff$denominator,
                                     genotype = "d17",
                                     sex = "F"
                                     )

pred_numerator_wt <- posterior_epred(fit1, newdata = pred_data_numerator_wt, re_formula = NA, cores = 1)
pred_numerator_d17 <- posterior_epred(fit1, newdata = pred_data_numerator_d17, re_formula = NA, cores = 1)
pred_denominator_wt <- posterior_epred(fit1, newdata = pred_data_denominator_wt, re_formula = NA, cores = 1)
pred_denominator_d17 <- posterior_epred(fit1, newdata = pred_data_denominator_d17, re_formula = NA, cores = 1)

pred_numerator <- pred_numerator_wt - pred_numerator_d17 
pred_denominator <- pred_denominator_wt - pred_denominator_d17 

comparisons_pred_numerator <- tidybayes::add_draws(comparisons_eff, pred_numerator, value = "pred_numerator")
comparisons_pred_denominator <- tidybayes::add_draws(comparisons_eff, pred_denominator, value = "pred_denominator")

comparisons_pred <- comparisons_pred_numerator %>% 
  inner_join(comparisons_pred_denominator, by = setdiff(names(comparisons_pred_numerator), "pred_numerator")) %>%
  mutate(log_ratio = log(eff_numerator) * pred_numerator - log(eff_denominator) * pred_denominator)


if(nrow(comparisons_pred) != nrow(comparisons_pred_numerator) || nrow(comparisons_pred) != nrow(comparisons_pred_denominator)) {
  stop("Bad join")
}

```

```{r}

qpcr_d17_averaged <- qpcr_d17_model %>% 
  group_by(genotype, primer_short) %>%
  summarise(avg_Cq = mean(Cq), sd_Cq = sd(Cq), sem_Cq = sd_Cq / n(), .groups = "drop")

qpcr_d17_averaged_comparisons <- comparisons_eff %>%
  inner_join(qpcr_d17_averaged, by = c("numerator" = "primer_short")) %>%
  rename(numerator_avg_Cq = avg_Cq, numerator_sem_Cq = sem_Cq) %>%
  inner_join(qpcr_d17_averaged, by = c("denominator" = "primer_short", "genotype")) %>%
  rename(denominator_avg_Cq = avg_Cq, denominator_sem_Cq = sem_Cq) #%>%
 # mutate(log_ratio = log(eff_numerator) * log(numerator_avg_Cq) - log(eff_denominator) * log(denominator_avg_Cq))

comparisons_obs <- qpcr_d17_averaged_comparisons %>%
  inner_join(qpcr_d17_averaged_comparisons, by = c("label", "comparison_id", "eff_numerator", "eff_denominator", "eff_label"), suffix = c("_d17","_wt")) %>%
  filter(genotype_d17 == "d17", genotype_wt == "wt") %>%
  mutate(log_ratio = log(eff_numerator) * (numerator_avg_Cq_wt - numerator_avg_Cq_d17) -  log(eff_denominator) * (denominator_avg_Cq_wt - denominator_avg_Cq_d17),
         sem_ratio = log(eff_numerator) * (numerator_sem_Cq_wt + numerator_sem_Cq_d17) -  log(eff_denominator) * (denominator_sem_Cq_wt + denominator_sem_Cq_d17))

```

```{r}
qpcr_d17_replicate_averaged <- qpcr_d17_model %>%
  group_by(run, animal_no, sex, genotype, primer_short) %>%
  summarise(avg_Cq = mean(Cq), .groups = "drop")


ratios_observed <- comparisons %>% 
  inner_join(qpcr_d17_replicate_averaged, by = c("numerator" = "primer_short")) %>%
  inner_join(qpcr_d17_replicate_averaged, by = c("denominator" = "primer_short", "run", "animal_no", "sex", "genotype"), suffix = c("_num", "_denom")) %>%
  mutate(Cq_denom_num = avg_Cq_denom - avg_Cq_num)

```

Below is a set of somewhat complex plots showing both A) the observed ratios of products within a biological sample for all of the comparisons modelled (left) and B) model-derived "ratios of ratios" between genotypes (right). 

Assuming equal PCR efficiency, the logarithm of ratios of two products can be approximated directly from the observed data by computing the difference in Cq numbers between the numerator and denominator products. In the left panel, each dot is a biological replicate and represents the difference between the mean Cq numbers of corresponding technical replicates. Since larger Cq implies lower expression, the larger the difference Cq (denominator) - Cq (numerator) the higher proportion of the numerator product was observed.

Using the statistical model, we can further compute the ratios of those ratios between genotypes in a hypothetical "noise free" experiment, after accounting for various sources of biological and technical variation. In the right panel we show the posterior credible intervals (95% - thin, 50% - thick) and means (points) for the ratio of ratios. Here ratio of ratios > 1 means that the d17 mice have larger numerator/denominator ratio than the wt mice. (note the log scale on the vertical axis)

```{r}

n_to_show <- 4

all_comparisons <- sort(unique(comparisons$label))

comparison_groups <- list()
comparison_groups[[1]] <- c()
for(comp in all_comparisons) {
  variants <- comparisons %>% filter(label == comp) %>% pull(label_variant)
  n_variants <- length(variants)
  last <- length(comparison_groups)
  if(last == 0 || length(comparison_groups[[last]]) + n_variants > n_to_show) {
    comparison_groups[[last + 1]] <- variants
  } else {
    comparison_groups[[last]] <- c(comparison_groups[[last]], variants)
  }
}

comparison_label_theme <- theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.3), axis.title.x = element_blank())


for(i in 1:length(comparison_groups)) {
  comparisons_to_show <- comparison_groups[[i]]
  
  # comparisons_obs_to_show <- comparisons_obs %>%
  #   filter(label_variant %in% comparisons_to_show, eff_label == "Neutral")
  
  nudge_width <- 0.2
  
  plot_pred <- comparisons_pred %>% 
    filter(label_variant %in% comparisons_to_show, eff_label == "Neutral") %>%
    ggplot(aes(x = label_variant, y = exp(log_ratio))) + 
    geom_hline(yintercept = 1, color = "blue") + 
    # geom_point(data = comparisons_obs_to_show, position = position_nudge(x = nudge_width), alpha = 0.8) +
    # geom_errorbar(data = comparisons_obs_to_show, aes(ymin = log_ratio - 1.96 * sem_ratio, ymax = log_ratio + 1.96 * sem_ratio), position = position_nudge(x = nudge_width), alpha = 0.8, width = nudge_width  / 2) +
    expand_limits(y = c(0.5, 2)) +
    stat_pointinterval() + 
    scale_y_log10("ratio d17/ratio wt") + comparison_label_theme

  plot_observed <- ratios_observed %>%
    filter(label_variant %in% comparisons_to_show) %>%
    ggplot(aes(x = genotype, y = Cq_denom_num)) + geom_boxplot(outlier.shape = NA) + geom_point(aes(color = sex, shape = sex), position = position_jitter(width = 0.3)) + facet_wrap(~label_variant, scales = "free_y", ncol = 2) + 
    scale_y_continuous("Cq (denominator) - Cq (numerator)")

  print((plot_observed | plot_pred) + plot_layout(widths = c(2,1)))
}


```

Computing the ratio of ratios requires us to guess a value for PCR efficiency. In the above plots we assumed both products have the same efficiency of 1.8. To check robustness, we take two additional cases - either the numerator or the denominator product have substantially larger efficiency - 1.5 vs 2.0. Many of the comparisons stay almost unaffected, but some do change the sign of the difference assuming different efficiency.

```{r}

comparisons_pred %>% 
  ggplot(aes(x = label_variant, y = exp(log_ratio), color = eff_label)) + 
  geom_hline(yintercept = 1, color = "blue") + 
  stat_pointinterval(position = position_dodge(width = 0.5)) + 
  scale_y_log10("ratio d17/ratio wt") + scale_color_brewer(type = "qual") +
  comparison_label_theme
```

# Mathematical model


Steibel et al.: Mixed models, normal error https://www.sciencedirect.com/science/article/pii/S0888754309000986
The same, but Poisson-LogNormal https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0071448



Below, assuming we have "noise free" $Cq$ values.

The ratio of the products $a,b$ in sample $i$ is:

$$
v_i = \frac{E_a^{S_a - Cq_{a,i}}}{E_b^{S_b - Cq_{b,i}}}
$$

Where $E_a, E_b$ are the PCR efficiencies and $S_a, S_b$ - the cycle that would be reached when amplifying a single molecule (similarly an abstract "noise free" value).

We are interested in 

$$
r = \frac{v_1}{v_2} \\
\log{r} = \log v_1 - \log v_2 = \left(  (\log E_a)(S_a - Cq_{a,1}) - (\log E_b)(S_b - Cq_{b,1}) \right) - 
\left(  (\log E_a)(S_a - Cq_{a,2}) - (\log E_b)(S_b - Cq_{b,2}) \right) = \\
(\log E_a)(Cq_{a,2} - Cq_{a,1}) - (\log E_b)(Cq_{b,2}-Cq_{b,1})
$$


$$
\log{r} = \log v_1 - \log v_2 = \left(  (\log E_a)(S_a - Cq_{a,1}) - (\log E_b)(S_b - Cq_{b,1}) \right) - \left(  (\log E_a)(S_a - Cq_{a,2}) - (\log E_b)(S_b - Cq_{b,2}) \right) = 
(\log E_a)(Cq_{a,2} - Cq_{a,1}) - (\log E_b)(Cq_{b,2}-Cq_{b,1})
$$
